app/services/report_generator_service.rb

```diff
@@ -0,0 +1,37 @@
+require "csv"
+
+class ReportGeneratorService
+  def initialize(filename)
+    @filename = filename
+  end
+
+  def call
+    CSV.open(filename, "w") do |csv|
+      csv << csv_headings
+      checks_with_errors.map do |check|
+        csv << csv_row_for_check(check)
+      end
+    end
+  end
+
+private
+
+  attr_reader :filename
+
+  def checks_with_errors
+    Link.includes(:checks)
+      .map(&:most_recent_check)
+      .compact
+      .select do |check|
+        check.completed? && check.has_errors?
+      end
```

Comment:
Not sure how concerned you are with performance here, but assuming there's a lot of checks per link (I can't remember quite how it works) you might want to switch this round to a query on check rather than links.

Reply:
okey doke 👍